File: LeetCode/python/SearchInsertPosition.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer}
    def searchInsert(self, nums, target):
        listSize = len(nums)
		
        if(listSize == 0):
            return 0
		
        if(target<nums[0]):
            return 0
		
        if(target > nums[listSize-1]):
            return listSize
		
        start = 0
        end = listSize
        mid = 0
        isFound = False
        while(start<end):
            mid = (start+end)//2
            # print("start ",start," end ",end," mid ",mid)
            if(nums[mid] == target):
                isFound = True
                break
			
            if(target<nums[mid]):
                end = mid
            else:
                start = mid+1
		
        if(not isFound and mid<listSize and end<listSize and
            nums[mid]<nums[end]):
            return mid+1
		
        return mid
		
#测试程序				
s = Solution()
l = [x for x in range(0,9,2)]

print(l)

for i in range(-2,12,1):
    print(i,s.searchInsert(l,i))

File: LeetCode/iter2/C++/SearchInsertPosition.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	int searchInsert(vector<int>& nums, int target) {
		int start = 0, end = nums.size() - 1;
		int candIndex = -1;
		while (start <= end)
		{
			int mid = start+(end-start) / 2;
			if (nums[mid] == target)
			{
				candIndex = mid;
				break;
			}
			else if (nums[mid] > target)
				end = mid - 1;
			else
				start = mid + 1;
		}

		if(candIndex == -1)
			candIndex = start;

		return candIndex;
	}
};

int main(void)
{
	vector<int> v;
	v.push_back(1);
	v.push_back(3);
	v.push_back(5);
	v.push_back(7);

	Solution s;
	for (int i = -3; i < 10; i++)
		cout << i<<" "<<s.searchInsert(v, i) << endl;


	return 0;
}